flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/AbstractSavepointSerializer.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.savepoint;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.checkpoint.SubtaskState;
+import org.apache.flink.runtime.checkpoint.TaskState;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.state.ChainedStateHandle;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.KeyGroupRangeOffsets;
+import org.apache.flink.runtime.state.KeyGroupsStateHandle;
+import org.apache.flink.runtime.state.OperatorStateHandle;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.runtime.state.filesystem.FileStateHandle;
+import org.apache.flink.runtime.state.memory.ByteStreamStateHandle;
+
+/**
+ * Abstract Serializer for {@link Savepoint} instances.
+ *
+ * <p>This is based on the {@link SavepointV1Serializer} of Flink 1.2.0 that
+ * makes sure no default Java serialization is used.
+ *
+ * <p>The abstract class allows to overwrite the serialization behaviour for
+ * {@link FileStateHandle} instances. This is the only practical difference
+ * between Flink 1.2.x and versions >= Flink 1.3.0.
+ *
+ * <p>This will probably be extended in ways that I cannot imagine at this point
+ * in time. If for whatever reason the abstract base class turns out to be a
+ * bad idea, feel free to change stuff. Right now, it's sole purpose is reducing
+ * code duplication between {@link SavepointV1Serializer} and {@link SavepointV2Serializer}.
+ */
+abstract class AbstractSavepointSerializer<T extends Savepoint> implements SavepointSerializer<T> {
```

Comment:
All of this is shared between V1 and V2 savepoints. I've moved it here and allowed to overwrite the file state handle serialization behaviour. @StefanRRichter, @StephanEwen The usefulness of this depends on how many changes we expect in the future. If we expect many, I think it is better to actually duplicate the code instead of moving it to an abstract class. For now, I'm slightly leaning towards not expecting many changes and hence moved it to the abstract class. In any case, moving back with this is a very trivial refactoring if you want the code to be rather copied.